stdlib/trains/trains.lua

```diff
@@ -0,0 +1,195 @@
+            -- Capture the rename
+            table.insert(renames, {old_id = id , new_id = derived_id, train = train })
+        end
+    end
+
+    -- Go over the captured renaming operations
+    for _, renaming in pairs(renames) do
+        -- Rename it in the registry
+        -- and dispatch a renamed event
+        Trains._registry[renaming.new_id] = renaming.train
+        table.remove_keys(Trains._registry, {renaming.old_id})
+
+        event_data = {
+            old_id = renaming.old_id,
+            new_id = renaming.new_id
+        }
+        Event.dispatch(Trains.on_train_renamed, event_data)
+    end
+end
+--- Event fired when a new locomotive has been created
+-- @tparam LuaEntity new_locomotive The new entity
+-- @return void
+function Trains.on_locomotive_created(new_locomotive)
+    train_id = Trains.get_train_id(new_locomotive.train)
+    if (Trains._registry[train_id] == nil) then
+        Trains._registry[train_id] = new_locomotive.train
+    end
+end
+
+
+-- Finds the distinct trains from an array of locomotive entities
+-- @tparam LuaEntity[] locomotives
+-- @return List of train details tables for the distinct trains
+function find_distinct_trains(locomotives)
+    train_data = {}
+    for _, engine in pairs(locomotives) do
+        local t = engine.train
+        local id = Trains.get_train_id(t)
+
+        local equals_id = function(x)
+            return x.id == id
+        end
+
+        local existing = table.filter(train_data, equals_id)
+
+        if #existing == 0 then
+            table.insert(train_data, {
+                train = t,
+                id = id
+            })
+        end
+    end
+
+    return train_data
+end
+
+-- Determines which locomotive in a train is the main one
+-- @tparam LuaTrain train
+-- @return LuaEntity the main locomotive
+function get_main_locomotive(train)
+    if train.valid and
+        train.locomotives and
+        (#train.locomotives.front_movers > 0 or #train.locomotives.back_movers > 0)
+    then
+        return train.locomotives.front_movers and train.locomotives.front_movers[1] or train.locomotives.back_movers[1]
+    end
+end
+
+-- Creates a registry of known trains
+-- @return table A mapping of train id to LuaTrain object
+function create_train_registry()
+    local registry = {}
+
+    local all_trains = Trains.find_filtered({surface_name = 1})
+    for _, trainInfo in pairs(all_trains) do
+        registry[tonumber(trainInfo.id)] = trainInfo.train
+    end
+
+    return registry
+end
+
+
+-- When developers  load this module, we need to
+-- attach some new events
+
+-- Filters events related to entity_type
+-- @tparam string event_parameter The event parameter to look inside to find the entity type
+-- @tparam string entity_type The entity type to filter events for
+-- @tparam callable callback The callback to invoke if the filter passes. The object defined in the event parameter is passed.
+function filter_event(event_parameter, entity_type, callback)

```

Comment:
I'd make this a local function. Without the local modifier, this function is part of the global namespace.